File: src/extras-kio/libdesktopexecparser.cpp
#define WORKAROUND_INNER_CLASS_DEFINITION_KIO__DesktopExecParser
#include <KService>
#include <QList>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QUrl>
#include <desktopexecparser.h>
#include "libdesktopexecparser.h"
#include "libdesktopexecparser.hxx"

KIO__DesktopExecParser* KIO__DesktopExecParser_new(const KService* service, const libqt_list /* of QUrl* */ urls) {
    QList<QUrl> urls_QList;
    urls_QList.reserve(urls.len);
    QUrl** urls_arr = static_cast<QUrl**>(urls.data);
    for (size_t i = 0; i < urls.len; ++i) {
        urls_QList.push_back(*(urls_arr[i]));
    }
    return new KIO::DesktopExecParser(*service, urls_QList);
}

void KIO__DesktopExecParser_SetUrlsAreTempFiles(KIO__DesktopExecParser* self, bool tempFiles) {
    self->setUrlsAreTempFiles(tempFiles);
}

void KIO__DesktopExecParser_SetSuggestedFileName(KIO__DesktopExecParser* self, const libqt_string suggestedFileName) {
    QString suggestedFileName_QString = QString::fromUtf8(suggestedFileName.data, suggestedFileName.len);
    self->setSuggestedFileName(suggestedFileName_QString);
}

libqt_list /* of libqt_string */ KIO__DesktopExecParser_ResultingArguments(const KIO__DesktopExecParser* self) {
    QList<QString> _ret = self->resultingArguments();
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_string* _arr = static_cast<libqt_string*>(malloc(sizeof(libqt_string) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        QString _lv_ret = _ret[i];
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _lv_b = _lv_ret.toUtf8();
        libqt_string _lv_str;
        _lv_str.len = _lv_b.length();
        _lv_str.data = static_cast<const char*>(malloc(_lv_str.len + 1));
        memcpy((void*)_lv_str.data, _lv_b.data(), _lv_str.len);
        ((char*)_lv_str.data)[_lv_str.len] = '\0';
        _arr[i] = _lv_str;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_string KIO__DesktopExecParser_ErrorMessage(const KIO__DesktopExecParser* self) {
    QString _ret = self->errorMessage();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_list /* of libqt_string */ KIO__DesktopExecParser_SupportedProtocols(const KService* service) {
    QList<QString> _ret = KIO::DesktopExecParser::supportedProtocols(*service);
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_string* _arr = static_cast<libqt_string*>(malloc(sizeof(libqt_string) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        QString _lv_ret = _ret[i];
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _lv_b = _lv_ret.toUtf8();
        libqt_string _lv_str;
        _lv_str.len = _lv_b.length();
        _lv_str.data = static_cast<const char*>(malloc(_lv_str.len + 1));
        memcpy((void*)_lv_str.data, _lv_b.data(), _lv_str.len);
        ((char*)_lv_str.data)[_lv_str.len] = '\0';
        _arr[i] = _lv_str;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

bool KIO__DesktopExecParser_IsProtocolInSupportedList(const QUrl* url, const libqt_list /* of libqt_string */ supportedProtocols) {
    QList<QString> supportedProtocols_QList;
    supportedProtocols_QList.reserve(supportedProtocols.len);
    libqt_string* supportedProtocols_arr = static_cast<libqt_string*>(supportedProtocols.data);
    for (size_t i = 0; i < supportedProtocols.len; ++i) {
        QString supportedProtocols_arr_i_QString = QString::fromUtf8(supportedProtocols_arr[i].data, supportedProtocols_arr[i].len);
        supportedProtocols_QList.push_back(supportedProtocols_arr_i_QString);
    }
    return KIO::DesktopExecParser::isProtocolInSupportedList(*url, supportedProtocols_QList);
}

bool KIO__DesktopExecParser_HasSchemeHandler(const QUrl* url) {
    return KIO::DesktopExecParser::hasSchemeHandler(*url);
}

libqt_string KIO__DesktopExecParser_ExecutableName(const libqt_string execLine) {
    QString execLine_QString = QString::fromUtf8(execLine.data, execLine.len);
    QString _ret = KIO::DesktopExecParser::executableName(execLine_QString);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string KIO__DesktopExecParser_ExecutablePath(const libqt_string execLine) {
    QString execLine_QString = QString::fromUtf8(execLine.data, execLine.len);
    QString _ret = KIO::DesktopExecParser::executablePath(execLine_QString);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void KIO__DesktopExecParser_Delete(KIO__DesktopExecParser* self) {
    delete self;
}

File: include_original/extras-kio/libdesktopexecparser.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");

/// https://api.kde.org/kio-desktopexecparser.html
pub const kio__desktopexecparser = struct {
    /// New constructs a new KIO::DesktopExecParser object.
    ///
    /// ``` service: QtC.KService, urls: []QtC.QUrl ```
    pub fn New(service: ?*anyopaque, urls: []QtC.QUrl) QtC.KIO__DesktopExecParser {
        const urls_list = qtc.libqt_list{
            .len = urls.len,
            .data = @ptrCast(urls.ptr),
        };

        return qtc.KIO__DesktopExecParser_new(@ptrCast(service), urls_list);
    }

    /// [Qt documentation](https://api.kde.org/kio-desktopexecparser.html#setUrlsAreTempFiles)
    ///
    /// ``` self: QtC.KIO__DesktopExecParser, tempFiles: bool ```
    pub fn SetUrlsAreTempFiles(self: ?*anyopaque, tempFiles: bool) void {
        qtc.KIO__DesktopExecParser_SetUrlsAreTempFiles(@ptrCast(self), tempFiles);
    }

    /// [Qt documentation](https://api.kde.org/kio-desktopexecparser.html#setSuggestedFileName)
    ///
    /// ``` self: QtC.KIO__DesktopExecParser, suggestedFileName: []const u8 ```
    pub fn SetSuggestedFileName(self: ?*anyopaque, suggestedFileName: []const u8) void {
        const suggestedFileName_str = qtc.libqt_string{
            .len = suggestedFileName.len,
            .data = suggestedFileName.ptr,
        };
        qtc.KIO__DesktopExecParser_SetSuggestedFileName(@ptrCast(self), suggestedFileName_str);
    }

    /// [Qt documentation](https://api.kde.org/kio-desktopexecparser.html#resultingArguments)
    ///
    /// ``` self: QtC.KIO__DesktopExecParser, allocator: std.mem.Allocator ```
    pub fn ResultingArguments(self: ?*anyopaque, allocator: std.mem.Allocator) [][]const u8 {
        const _arr: qtc.libqt_list = qtc.KIO__DesktopExecParser_ResultingArguments(@ptrCast(self));
        const _str: [*]qtc.libqt_string = @ptrCast(@alignCast(_arr.data));
        defer {
            for (0.._arr.len) |i| {
                qtc.libqt_string_free(@ptrCast(&_str[i]));
            }
            qtc.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc([]const u8, _arr.len) catch @panic("kio::desktopexecparser.ResultingArguments: Memory allocation failed");
        for (0.._arr.len) |i| {
            const _data = _str[i];
            const _buf = allocator.alloc(u8, _data.len) catch @panic("kio::desktopexecparser.ResultingArguments: Memory allocation failed");
            @memcpy(_buf, _data.data[0.._data.len]);
            _ret[i] = _buf;
        }
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kio-desktopexecparser.html#errorMessage)
    ///
    /// ``` self: QtC.KIO__DesktopExecParser, allocator: std.mem.Allocator ```
    pub fn ErrorMessage(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.KIO__DesktopExecParser_ErrorMessage(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kio::desktopexecparser.ErrorMessage: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kio-desktopexecparser.html#supportedProtocols)
    ///
    /// ``` service: QtC.KService, allocator: std.mem.Allocator ```
    pub fn SupportedProtocols(service: ?*anyopaque, allocator: std.mem.Allocator) [][]const u8 {
        const _arr: qtc.libqt_list = qtc.KIO__DesktopExecParser_SupportedProtocols(@ptrCast(service));
        const _str: [*]qtc.libqt_string = @ptrCast(@alignCast(_arr.data));
        defer {
            for (0.._arr.len) |i| {
                qtc.libqt_string_free(@ptrCast(&_str[i]));
            }
            qtc.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc([]const u8, _arr.len) catch @panic("kio::desktopexecparser.SupportedProtocols: Memory allocation failed");
        for (0.._arr.len) |i| {
            const _data = _str[i];
            const _buf = allocator.alloc(u8, _data.len) catch @panic("kio::desktopexecparser.SupportedProtocols: Memory allocation failed");
            @memcpy(_buf, _data.data[0.._data.len]);
            _ret[i] = _buf;
        }
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kio-desktopexecparser.html#isProtocolInSupportedList)
    ///
    /// ``` url: QtC.QUrl, supportedProtocols: [][]const u8, allocator: std.mem.Allocator ```
    pub fn IsProtocolInSupportedList(url: ?*anyopaque, supportedProtocols: [][]const u8, allocator: std.mem.Allocator) bool {
        var supportedProtocols_arr = allocator.alloc(qtc.libqt_string, supportedProtocols.len) catch @panic("kio::desktopexecparser.IsProtocolInSupportedList: Memory allocation failed");
        defer allocator.free(supportedProtocols_arr);
        for (supportedProtocols, 0..supportedProtocols.len) |item, i| {
            supportedProtocols_arr[i] = .{
                .len = item.len,
                .data = item.ptr,
            };
        }
        const supportedProtocols_list = qtc.libqt_list{
            .len = supportedProtocols.len,
            .data = supportedProtocols_arr.ptr,
        };
        return qtc.KIO__DesktopExecParser_IsProtocolInSupportedList(@ptrCast(url), supportedProtocols_list);
    }

    /// [Qt documentation](https://api.kde.org/kio-desktopexecparser.html#hasSchemeHandler)
    ///
    /// ``` url: QtC.QUrl ```
    pub fn HasSchemeHandler(url: ?*anyopaque) bool {
        return qtc.KIO__DesktopExecParser_HasSchemeHandler(@ptrCast(url));
    }

    /// [Qt documentation](https://api.kde.org/kio-desktopexecparser.html#executableName)
    ///
    /// ``` execLine: []const u8, allocator: std.mem.Allocator ```
    pub fn ExecutableName(execLine: []const u8, allocator: std.mem.Allocator) []const u8 {
        const execLine_str = qtc.libqt_string{
            .len = execLine.len,
            .data = execLine.ptr,
        };
        const _str = qtc.KIO__DesktopExecParser_ExecutableName(execLine_str);
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kio::desktopexecparser.ExecutableName: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kio-desktopexecparser.html#executablePath)
    ///
    /// ``` execLine: []const u8, allocator: std.mem.Allocator ```
    pub fn ExecutablePath(execLine: []const u8, allocator: std.mem.Allocator) []const u8 {
        const execLine_str = qtc.libqt_string{
            .len = execLine.len,
            .data = execLine.ptr,
        };
        const _str = qtc.KIO__DesktopExecParser_ExecutablePath(execLine_str);
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kio::desktopexecparser.ExecutablePath: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.KIO__DesktopExecParser ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.KIO__DesktopExecParser_Delete(@ptrCast(self));
    }
};
